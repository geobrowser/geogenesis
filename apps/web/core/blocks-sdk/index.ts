export * as TableBlockSdk from './table';
