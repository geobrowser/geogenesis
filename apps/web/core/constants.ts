export const ZERO_WIDTH_SPACE = '\u200b';

export const PLACEHOLDER_IMAGES = {
  'San Francisco': 'https://cdn.discordapp.com/attachments/1010259815395754147/1047945562336534588/SF-image.png',
  Health: 'https://images.unsplash.com/photo-1535914254981-b5012eebbd15',
  Values: 'https://images.unsplash.com/photo-1637008336770-b95d637fd5fa',
};

export const ALL_SPACES_IMAGE = 'ipfs://QmQXJYrbJJZcukgkzk8C71nQL1V8ND9TQjrtP4sKWjYPFH';

export const ROOT_SPACE_IMAGE =
  'https://api.thegraph.com/ipfs/api/v0/cat?arg=QmU3TRxyXGKbgCFPFVddJmRnxHQ6ojnppumU3zpPUjJ4WK';

export const DEFAULT_OPENGRAPH_IMAGE = 'https://www.geobrowser.io/static/geo-social-image-v2.png';

export const DEFAULT_OPENGRAPH_DESCRIPTION =
  "Browse and organize the world's public knowledge and information in a decentralized way.";
