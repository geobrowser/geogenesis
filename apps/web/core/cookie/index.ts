export * as Cookie from './cookie';
export const HAS_DISMISSED_PERSONAL_SPACE_ONBOARDING_KEY = 'hasDismissedPersonalSpaceOnboarding';
export const WALLET_ADDRESS = 'walletAddress';
