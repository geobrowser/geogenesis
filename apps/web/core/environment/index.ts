export type { AppConfig } from './environment';
export * as Environment from './environment';
