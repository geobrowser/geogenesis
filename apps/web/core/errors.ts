export class TransactionWriteFailedError extends Error {
  readonly _tag = 'TransactionWriteFailedError';
}

export class IpfsUploadError extends Error {
  readonly _tag = 'IpfsUploadError';
}
