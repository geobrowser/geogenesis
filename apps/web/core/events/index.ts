export { useEditEvents } from './edit-events';
export type { EditEvent } from './edit-events';
