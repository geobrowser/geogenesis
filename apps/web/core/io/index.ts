export { MockNetworkData } from './mocks';
export { Storage } from './storage';
export * as Subgraph from './subgraph';
export * as Publish from './publish';
