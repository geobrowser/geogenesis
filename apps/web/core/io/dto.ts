import { PLACEHOLDER_SPACE_IMAGE } from '../constants';
import { Value } from '../types';
import { Entities } from '../utils/entity';
import { EntityDto } from './dto/entities';
import { SpaceConfigEntity } from './dto/spaces';
import { EntityId, SubstreamEntity, SubstreamTriple } from './schema';

function extractValue(networkTriple: SubstreamTriple): Value {
  switch (networkTriple.valueType) {
    case 'TEXT':
      return { type: 'TEXT', value: networkTriple.textValue };
    case 'ENTITY': {
      return {
        type: 'ENTITY',
        value: networkTriple.entityValue.id,
        name: networkTriple.entityValue.name,
      };
    }
    case 'TIME':
      return { type: 'TIME', value: networkTriple.textValue };
    case 'URI':
      return { type: 'URI', value: networkTriple.textValue };
  }
}

export function TripleDto(triple: SubstreamTriple) {
  return {
    entityId: triple.entity.id,
    entityName: triple.entity.name,
    attributeId: triple.attribute.id,
    attributeName: triple.attribute.name,
    value: extractValue(triple),
    space: triple.space.id,
  };
}

export function SpaceMetadataDto(spaceId: string, metadata: SubstreamEntity | undefined | null): SpaceConfigEntity {
  const entity = metadata ? EntityDto(metadata) : null;

  const spaceConfigWithImage: SpaceConfigEntity = entity
    ? {
        ...entity,
        spaceId: spaceId,
        image: Entities.avatar(entity.relationsOut) ?? Entities.cover(entity.relationsOut) ?? PLACEHOLDER_SPACE_IMAGE,
      }
    : {
        id: EntityId(''),
        spaceId: spaceId,
        name: null,
        description: null,
        image: PLACEHOLDER_SPACE_IMAGE,
        triples: [],
        types: [],
        nameTripleSpaces: [],
        relationsOut: [],
      };

  return spaceConfigWithImage;
}
