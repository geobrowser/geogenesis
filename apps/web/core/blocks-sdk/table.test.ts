import { SYSTEM_IDS } from '@geogenesis/sdk';
import { describe, expect, it } from 'vitest';

import { MockNetworkData } from '~/core/io';

import {
  createFiltersFromGraphQLString,
  createGraphQLStringFromFilters,
  createGraphQLStringFromFiltersV2,
} from './table';

describe('TableBlock SDK', () => {
  /**
   * There are several combinations of filters that can be applied to a table block.
   * 1. String field with a string value
   * 2. Entity field with an entity ID as the value
   * 3. String field targeting the Name column
   * 4. A null type id
   *
   * These four combinations can also be used together with an "and" filter
   */
  it('Builds a graphql query from table block filters', () => {
    const stringFilter = createGraphQLStringFromFilters(
      [
        {
          columnId: 'type',
          value: 'Value 1',
          valueType: 'TEXT',
        },
      ],
      'type-id'
    );

    expect(stringFilter).toEqual(
      `{typeIds_contains_nocase: ["type-id"], entityOf_: {attribute: "type", stringValue_starts_with_nocase: "Value 1"}}`
    );

    const entityFilter = createGraphQLStringFromFilters(
      [
        {
          columnId: 'type',
          value: 'id 1',
          valueType: 'ENTITY',
        },
      ],
      'type-id'
    );

    expect(entityFilter).toEqual(
      `{typeIds_contains_nocase: ["type-id"], entityOf_: {attribute: "type", entityValue: "id 1"}}`
    );

    const nameFilter = createGraphQLStringFromFilters(
      [
        {
          columnId: SYSTEM_IDS.NAME,
          value: 'id 1',
          valueType: 'TEXT',
        },
      ],
      'type-id'
    );

    expect(nameFilter).toEqual(`{typeIds_contains_nocase: ["type-id"], name_starts_with_nocase: "id 1"}`);

    const andFilter = createGraphQLStringFromFilters(
      [
        {
          columnId: 'type',
          value: 'Value 1',
          valueType: 'TEXT',
        },
        {
          columnId: 'type',
          value: 'id 1',
          valueType: 'ENTITY',
        },
        {
          columnId: SYSTEM_IDS.NAME,
          value: 'id 1',
          valueType: 'TEXT',
        },
      ],
      'type-id'
    );

    expect(andFilter).toEqual(
      `{and: [{typeIds_contains_nocase: ["type-id"]}, {entityOf_: {attribute: "type", stringValue_starts_with_nocase: "Value 1"}}, {entityOf_: {attribute: "type", entityValue: "id 1"}}, {name_starts_with_nocase: "id 1"}]}`
    );

    const spaceFilter = createGraphQLStringFromFilters(
      [
        {
          columnId: SYSTEM_IDS.SPACE,
          valueType: 'TEXT',
          value: '0x0000000000000000000000000000000000000000',
        },
      ],
      'type-id'
    );

    expect(spaceFilter).toEqual(
      `{typeIds_contains_nocase: ["type-id"], entityOf_: {space: "0x0000000000000000000000000000000000000000"}}`
    );

    const nullTypeIdFilter = createGraphQLStringFromFilters([], null);

    expect(nullTypeIdFilter).toEqual('');
  });

  it('Builds the TableBlockStore filters data structure from a graphql string', async () => {
    /**
     * There are several combinations of filters that can be applied to a table block.
     * 1. String field with a string value
     * 2. Entity field with an entity ID as the value
     * 3. String field targeting the Name column
     * 4. A null type id
     *
     * These four combinations can also be used together with an "and" filter
     */
    const stringFilter = await createFiltersFromGraphQLString(
      `{typeIds_contains_nocase: ["type-id"], entityOf_: {attribute: "type", stringValue_starts_with_nocase: "Value 1"}}`,
      async () => {
        return {
          id: 'type',
          triples: [MockNetworkData.makeStubTriple('Types')],
          name: 'Types',
          description: '',
          types: [],
        };
      }
    );

    expect(stringFilter).toEqual([
      {
        columnId: 'type',
        value: 'Value 1',
        valueType: 'string',
        valueName: null,
      },
    ]);

    const entityFilter = await createFiltersFromGraphQLString(
      `{typeIds_contains_nocase: ["type-id"], entityOf_: {attribute: "type", entityValue: "id 1"}}`,
      async () => {
        return {
          id: 'id 1',
          triples: [MockNetworkData.makeStubTriple('Types')],
          name: 'Entity Name',
          description: '',
          types: [],
        };
      }
    );

    expect(entityFilter).toEqual([{ columnId: 'type', valueName: 'Entity Name', value: 'id 1', valueType: 'entity' }]);

    const nameFilter = await createFiltersFromGraphQLString(
      `{typeIds_contains_nocase: ["type-id"], name_starts_with_nocase: "id 1"}`,
      async () => {
        return {
          id: 'type',
          triples: [MockNetworkData.makeStubTriple('Name')],
          name: 'Name',
          description: '',
          types: [],
        };
      }
    );

    expect(nameFilter).toEqual([
      {
        columnId: SYSTEM_IDS.NAME,
        value: 'id 1',
        valueType: 'string',
        valueName: null,
      },
    ]);

    const andFilter = await createFiltersFromGraphQLString(
      `{and: [{typeIds_contains_nocase: ["type-id"]}, {entityOf_: {attribute: "type", stringValue_starts_with_nocase: "Value 1"}}, {entityOf_: {attribute: "type", entityValue: "id 1"}}, {name_starts_with_nocase: "id 1"}]}`,
      async () => {
        return {
          id: 'id 1',
          triples: [MockNetworkData.makeStubTriple('Types')],
          name: 'Entity Name',
          description: '',
          types: [],
        };
      }
    );

    expect(andFilter).toEqual([
      {
        columnId: SYSTEM_IDS.NAME,
        value: 'id 1',
        valueType: 'string',
        valueName: null,
      },
      {
        columnId: 'type',
        value: 'id 1',
        valueType: 'entity',
        valueName: 'Entity Name',
      },
      {
        columnId: 'type',
        value: 'Value 1',
        valueType: 'string',
        valueName: null,
      },
    ]);

    const spaceFilter = await createFiltersFromGraphQLString(
      `{typeIds_contains_nocase: ["type-id"], entityOf_: {space: "0x0000000000000000000000000000000000000000"}}`,
      async () => {
        return {
          id: 'id 1',
          triples: [MockNetworkData.makeStubTriple('Types')],
          name: 'Entity Name',
          description: '',
          types: [],
        };
      }
    );

    expect(spaceFilter).toEqual([
      {
        columnId: SYSTEM_IDS.SPACE,
        value: '0x0000000000000000000000000000000000000000',
        valueType: 'string',
        valueName: null,
      },
    ]);
  });

  it('Builds a graphql query from table block filters for the postgraphile-based substreams API', () => {
    const stringFilter = createGraphQLStringFromFiltersV2(
      [
        {
          columnId: 'type',
          value: 'Value 1',
          valueType: 'TEXT',
        },
      ],
      'type-id'
    );

    expect(stringFilter).toMatchSnapshot();

    const entityFilter = createGraphQLStringFromFiltersV2(
      [
        {
          columnId: 'type',
          value: 'id 1',
          valueType: 'ENTITY',
        },
      ],
      'type-id'
    );

    expect(entityFilter).toMatchSnapshot();

    const nameFilter = createGraphQLStringFromFiltersV2(
      [
        {
          columnId: SYSTEM_IDS.NAME,
          value: 'id 1',
          valueType: 'TEXT',
        },
      ],
      'type-id'
    );

    expect(nameFilter).toMatchSnapshot();

    const andFilter = createGraphQLStringFromFiltersV2(
      [
        {
          columnId: 'type',
          value: 'Value 1',
          valueType: 'TEXT',
        },
        {
          columnId: 'type',
          value: 'id 1',
          valueType: 'ENTITY',
        },
        {
          columnId: SYSTEM_IDS.NAME,
          value: 'id 1',
          valueType: 'TEXT',
        },
      ],
      'type-id'
    );

    expect(andFilter).toMatchSnapshot();

    const spaceFilter = createGraphQLStringFromFiltersV2(
      [
        {
          columnId: SYSTEM_IDS.SPACE,
          valueType: 'TEXT',
          value: '0x0000000000000000000000000000000000000000',
        },
      ],
      'type-id'
    );

    expect(spaceFilter).toMatchSnapshot();

    const nullTypeIdFilter = createGraphQLStringFromFiltersV2([], null);

    expect(nullTypeIdFilter).toEqual('');
  });
});
