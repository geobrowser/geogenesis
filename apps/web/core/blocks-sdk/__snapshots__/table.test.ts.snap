// Vitest Snapshot v1

exports[`TableBlock SDK > Builds a graphql query from table block filters for the postgraphile-based substreams API 1`] = `"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, triples: { some: { attributeId: { equalTo: \\"type\\" }, stringValue: { equalToInsensitive: \\"Value 1\\"}, isStale: { equalTo: false } } }}"`;

exports[`TableBlock SDK > Builds a graphql query from table block filters for the postgraphile-based substreams API 2`] = `"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, triples: { some: { attributeId: { equalTo: \\"type\\" }, entityValueId: { equalTo: \\"id 1\\"}, isStale: { equalTo: false } } }}"`;

exports[`TableBlock SDK > Builds a graphql query from table block filters for the postgraphile-based substreams API 3`] = `"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, name: { startsWithInsensitive: \\"id 1\\" }}"`;

exports[`TableBlock SDK > Builds a graphql query from table block filters for the postgraphile-based substreams API 4`] = `"{ and: [{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } } } { triples: { some: { attributeId: { equalTo: \\"type\\" }, stringValue: { equalToInsensitive: \\"Value 1\\"}, isStale: { equalTo: false } } } }, { triples: { some: { attributeId: { equalTo: \\"type\\" }, entityValueId: { equalTo: \\"id 1\\"}, isStale: { equalTo: false } } } }, { name: { startsWithInsensitive: \\"id 1\\" } }] }"`;

exports[`TableBlock SDK > Builds a graphql query from table block filters for the postgraphile-based substreams API 5`] = `
"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, triples: {
          some: {
            spaceId: { equalTo: \\"0x0000000000000000000000000000000000000000\\" },
            isStale: { equalTo: false }
          }
        }}"
`;

exports[`TableBlock SDK > Substream: Builds a graphql query from table block filters 1`] = `"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, triples: { some: { attributeId: { equalTo: \\"type\\" }, stringValue: { equalToInsensitive: \\"Value 1\\"}, isStale: { equalTo: false } } }}"`;

exports[`TableBlock SDK > Substream: Builds a graphql query from table block filters 2`] = `"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, triples: { some: { attributeId: { equalTo: \\"type\\" }, entityValueId: { equalTo: \\"id 1\\"}, isStale: { equalTo: false } } }}"`;

exports[`TableBlock SDK > Substream: Builds a graphql query from table block filters 3`] = `"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, name: { startsWithInsensitive: \\"id 1\\" }}"`;

exports[`TableBlock SDK > Substream: Builds a graphql query from table block filters 4`] = `"{ and: [{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } } } { triples: { some: { attributeId: { equalTo: \\"type\\" }, stringValue: { equalToInsensitive: \\"Value 1\\"}, isStale: { equalTo: false } } } }, { triples: { some: { attributeId: { equalTo: \\"type\\" }, entityValueId: { equalTo: \\"id 1\\"}, isStale: { equalTo: false } } } }, { name: { startsWithInsensitive: \\"id 1\\" } }] }"`;

exports[`TableBlock SDK > Substream: Builds a graphql query from table block filters 5`] = `
"{ geoEntityTypesByEntityId: { some: { typeId: { equalTo: \\"type-id\\" } } }, triples: {
          some: {
            spaceId: { equalTo: \\"0x0000000000000000000000000000000000000000\\" },
            isStale: { equalTo: false }
          }
        }}"
`;
